### DISCUSSION TITLE
Writing to file with images concurrently

### DISCUSSION THREAD

**Discussion by WebdevMerlion** [exceljs/exceljs#2498] [Q&A]
Ok, guys. I have a json includes a remote url of image. trying to get images and add to worksheet by Promise.all http.get. So code is:

` 

      const getImage = (url, book, sheet, sizeOf, row, column) => new Promise((resolve, reject) => {

        const request = https.get(url, response => {
            if(response.statusCode === 200){

                let chunks = [];

                const imageColumn = sheet.getColumn('image')
                const imageRow = sheet.getRow(row + 1)

                response.on('data', chunk => {
                    chunks.push(chunk)
                }).on('end', () => {
                    const buffer = Buffer.concat(chunks)
                    const dimension = sizeOf(buffer)

                    imageColumn.width = dimension.width
                    imageRow.height = dimension.height

                    const imageId = book.addImage({
                        buffer: buffer,
                        extension: 'png',
                    });

                    sheet.addImage(imageId, {
                        tl: { col: column, row: row },
                        br: { col: column + 1, row: row + 1 },
                        editAs: "oneCell"
                    });

                    resolve(book)
                })
            }

        })

        request.on('error', (err) => reject(err))
    })

    sheet.columns = Object.keys(header).map((index) => ({Header: header[index], key: index}))
    let promises = []

    req.body.forEach((row, index) => {
        sheet.addRow(row)
        row.image && /\.(jpg|png|gif)/.test(row.image) && promises.push(getImage(url.parse(row.image), workbook, sheet, sizeOf, index, imageColumn))
    })

    promises.length
        && Promise.all(promises).then(book => book.pop().xlsx.write(res).then(() => res.end()), err => console.log(err))
        || writer.write(res).then(() => res.end())`

And i have an 504 timeout on a rows more than 500. Are image operation with excel blocking my event-loop, or buffer operations? How can i get all images and then add it to book? Maybe i should use a stremaing i/o to write file in child processes concurrently?

**Comment by Siemienik** [comment]
Hello @WebdevMerlion , I've just quick read Your code without deep analyzing, I noted only:
AD. 
```
sheet.columns = Object.keys(header).map((index) => ({Header: header[index], key: index}))
```
should be `header` instead of `Header`


I hope it is helpful for You ;) I'll come back here when I will have more time.


**EDIT**: *maybe It is a bug, so I put this label to keep issue open*

**Comment by WebdevMerlion** [comment]
Hi, @Siemienik. header it's just first line of my json `const header = req.body[0]`. Have no trouble with sheet columns init. Same trouble when using async.queue
